staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,436 @@
+		// This might be a built-in command, external plugin, etc.
+		return args, nil
+	}
+
+	rootCmd.AddCommand(aliasArgs.command)
+
+	foundAliasCmd, _, err := rootCmd.Find([]string{commandName})
+	if err != nil {
+		return args, nil
+	}
+
+	// This function triggers merging the persistent flags in the parent commands.
+	_ = foundAliasCmd.InheritedFlags()
+
+	allShorthands := make(map[string]struct{})
+	foundAliasCmd.Flags().VisitAll(func(flag *pflag.Flag) {
+		if flag.Shorthand != "" {
+			allShorthands[flag.Shorthand] = struct{}{}
+		}
+	})
+
+	for _, fl := range aliasArgs.flags {
+		existingFlag := foundAliasCmd.Flag(fl.Name)
+		if existingFlag == nil {
+			return args, fmt.Errorf("invalid alias flag %s in alias %s", fl.Name, args[0])
+		}
+		if searchInArgs(existingFlag.Name, existingFlag.Shorthand, allShorthands, args) {
+			// Don't modify the value implicitly, if it is passed in args explicitly
+			continue
+		}
+		err = foundAliasCmd.Flags().Set(fl.Name, fl.Default)
+		if err != nil {
+			return args, fmt.Errorf("could not apply value %s to flag %s in alias %s err: %w", fl.Default, fl.Name, args[0], err)
+		}
+	}
+
+	if len(aliasArgs.prependArgs) > 0 {
+		// prependArgs defined in kuberc should be inserted after the alias name.
+		if commandIndex+1 >= len(args) {
+			// command is the last item, we simply append just like appendArgs
+			args = append(args, aliasArgs.prependArgs...)
+		} else {
+			args = append(args[:commandIndex+1], append(aliasArgs.prependArgs, args[commandIndex+1:]...)...)
+		}
+	}
+	if len(aliasArgs.appendArgs) > 0 {
+		// appendArgs defined in kuberc should be appended to actual args.
+		args = append(args, aliasArgs.appendArgs...)
+	}
+	// Cobra (command.go#L1078) appends only root command's args into the actual args and ignores the others.
+	// We are appending the additional args defined in kuberc in here and
+	// expect that it will be passed along to the actual command.
+	rootCmd.SetArgs(args[1:])
+	return args, nil
+}
+
+// DefaultGetPreferences returns KubeRCConfiguration.
+// If users sets kuberc file explicitly in --kuberc flag, it has the highest
+// priority. If not specified, it looks for in KUBERC environment variable.
+// If KUBERC is also not set, it falls back to default .kuberc file at the same location
+// where kubeconfig's defaults are residing in.
+func DefaultGetPreferences(kuberc string) (*config.Preference, error) {
+	kubeRCFile := RecommendedKubeRCFile
+	explicitly := false
+	if kuberc != "" {
+		kubeRCFile = kuberc
+		explicitly = true
+	}
+
+	if kubeRCFile == "" && os.Getenv("KUBERC") != "" {
+		kubeRCFile = os.Getenv("KUBERC")
+		explicitly = true
+	}
+
+	kubeRCBytes, err := os.ReadFile(kubeRCFile)
+	if err != nil {
+		if os.IsNotExist(err) && !explicitly {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	obj, gvk, err := codecs.UniversalDecoder().Decode(kubeRCBytes, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not be decoded gvk %s, err: %w", gvk, err)
+	}
```

Comment:
we need to think through whether a decoding errors (either because the GVK wasn't recognized, or because strict decoding saw an unknown field) should be treated as fatal by kubectl when kuberc wasn't explicitly requested

Scenarios:
1. In 1.33 or whatever, we promote this feature to beta, enable it by default in kubectl, and promote the config version to beta. in 1.34 or whatever we promote the config file to v1. Then, 1.33 kubectls will start crashing when run on a machine with a v1 kuberc present. That seems... bad.
2. In some future version, we add a new top-level field to the preferences ("color preferences" or something random). Then, previous kubectls start crashing when run on a machine that uses this preference.

Possibilities:
* Only treat decode errors as fatal when kuberc is explicitly requested (via flag? via envvar?), and warn otherwise
* Try strict decoding and fall back to lenient decoding... if lenient decoding succeeds, warn but continue (we've done this in other places)

Reply:
> Try strict decoding and fall back to lenient decoding... if lenient decoding succeeds, warn but continue (we've done this in other places)

I tried to apply this fallback mechanism. Please let me know it looks correct. Thanks.

Reply:
@liggitt I believe that I have applied all your suggestions. PTAL?.